### PR TITLE
Semaphore API

### PR DESCRIPTION
**Pull request by Shishqa**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

### CODE REVIEW

File: ydb/examples/mutex.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        .await?;
+
+    session
+        .create_semaphore("my-resource".to_string(), ydb::SemaphoreLimit::Mutex, None)
+        .await?;
+
+    println!("done");
+
+    let mut handles: Vec<JoinHandle<()>> = vec![];
+    for _ in range(0, 4) {
```
Comment (rekby) [inline ydb/examples/mutex.rs:62]: why range instead of for _ in 0..4?

File: ydb/src/client_coordination/list_types.rs [hunk 2]
```diff
@@ -100,3 +99,51 @@ impl From<RawDescribeNodeResult> for NodeDescription {
         }
     }
 }
+
+#[derive(Debug, serde::Serialize)]
+#[allow(dead_code)]
+pub struct SemaphoreSession {
```
Comment (rekby) [inline ydb/src/client_coordination/list_types.rs:105]: Need to move wrappers of grpc messages to grpc_wrapper

**Review by rekby — COMMENTED** [review]
great work!

File: ydb/examples/mutex.rs (hunk 1 above)
Comment (Shishqa) [reply ydb/examples/mutex.rs]: fixed

File: ydb/src/client_coordination/list_types.rs (hunk 2 above)
Comment (Shishqa) [reply ydb/src/client_coordination/list_types.rs]: fixed

File: ydb/src/client_coordination/session/coordination_session.rs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+};
+
+use super::{
+    controller::RequestController, create_options::SemaphoreLimit, describe_options::WatchOptions,
+    lease::Lease,
+};
+
+#[derive(Clone)]
+struct MethodControllers {
+    pub create_semaphore: Arc<RequestController<RawCreateSemaphoreResult>>,
```
Comment (Shishqa) [inline ydb/src/client_coordination/session/coordination_session.rs:43]: make private

File: ydb/src/client_coordination/session/acquire_options.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        }
+    }
+}
+
+#[derive(Builder, Clone)]
+#[builder(build_fn(error = "errors::YdbError"))]
+#[allow(dead_code)]
+pub struct AcquireOptions {
+    #[builder(setter(strip_option), default)]
+    pub data: Option<Vec<u8>>,
```
Comment (Shishqa) [inline ydb/src/client_coordination/session/acquire_options.rs:27]: make `pub` or `pub(crate)` consistent with the rest of the sdk

File: ydb/src/client_coordination/session/coordination_session.rs [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+            Err(YdbError::Custom("failed to acquire semaphore".to_string()))
+        }
+    }
+
+    pub fn id(&self) -> u64 {
+        self.id
+    }
+
+    pub fn client(&self) -> CoordinationClient {
+        unimplemented!()
```
Comment (Shishqa) [inline ydb/src/client_coordination/session/coordination_session.rs:283]: remove

File: ydb/src/client_coordination/session/acquire_options.rs [hunk 6]
```diff
@@ -0,0 +1,34 @@
+use crate::errors;
+use derive_builder::Builder;
+use std::time::Duration;
+
+#[derive(Debug)]
+pub enum AcquireCount {
+    Single,
+    Exclusive,
+    Custom(u64),
+}
```
Comment (Shishqa) [inline ydb/src/client_coordination/session/acquire_options.rs:10]: different methods for different options

File: ydb/src/client_coordination/session/create_options.rs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub enum SemaphoreLimit {
+    Mutex,
+    Unbounded,
+    Custom(u64),
+}
```
Comment (Shishqa) [inline ydb/src/client_coordination/session/create_options.rs:7]: make different methods

File: ydb/src/client_coordination/session/coordination_session.rs [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+            None => Err(YdbError::Custom("channel closed".to_string())),
+        }
+    }
+
+    pub async fn watch_semaphore(
+        &self,
+        _name: String,
+        _options: WatchOptions,
+    ) -> YdbResult<mpsc::Receiver<SemaphoreDescription>> {
+        unimplemented!()
```
Comment (Shishqa) [inline ydb/src/client_coordination/session/coordination_session.rs:207]: remove for now

File: ydb/src/client_coordination/session/coordination_session.rs [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+        let mut stream = coordination_service.session(session_start_request).await?;
+        let start_response = stream.receive::<RawSessionResponse>().await?;
+
+        let session_response: SessionStarted;
+        if let RawSessionResponse::SessionStarted(response) = start_response {
+            session_response = response;
+        } else {
+            return Err(YdbError::Custom("unexpected session answer".to_string()));
+        }
+        println!("session started! {:?}", session_response);
```
Comment (Shishqa) [inline ydb/src/client_coordination/session/coordination_session.rs:102]: `trace!`

File: ydb/src/coordination_test.rs [hunk 10]
```diff
@@ -61,3 +63,139 @@ async fn create_delete_node_test() -> YdbResult<()> {
...
+        ) => {
+            let description = result.unwrap();
+            assert_eq!(description.ephemeral, ephemeral);
+            assert_eq!(description.owners.len(), 1);
+            assert_eq!(description.owners[0].data, vec![i]);
+            assert_eq!(description.owners[0].session_id, session.id());
+        },
+    }
+}
+
```
Comment (Shishqa) [inline ydb/src/coordination_test.rs:106]: /+ describe test
/+ update test

**Review by rekby — APPROVED** [review]

**Pull request merged**